lib/vault.ts

```diff
@@ -191,6 +197,164 @@ export async function listAllSubmissions(
+
+      if (response.Responses?.Vault.length) {
+        accumulatedResponses = accumulatedResponses.concat(
+          response.Responses.Vault.map((item) => {
+            return {
+              formID: item.FormID.S,
+              submissionID: item.SubmissionID.S,
+              formSubmission: item.FormSubmission.S,
+              confirmationCode: item.ConfirmationCode.S,
+              status: item.Status.S as VaultStatus,
+              securityAttribute: item.SecurityAttribute.S,
+              name: item.Name.S,
+              createdAt: item.CreatedAt.N,
+              lastDownloadedBy: item.LastDownloadedBy?.S ?? null,
+              confirmedAt: item.ConfirmTimestamp?.N ?? null,
+              downloadedAt: item.DownloadedAt?.N ?? null,
+              removedAt: item.RemovalDate?.N ?? null,
+            } as unknown as VaultSubmission;
+          })
+        );
+      }
+
+      // If there are unprocessed keys, we need to make another request
+      keys = response.UnprocessedKeys?.Vault?.Keys || [];
+    }
+
+    // Log each response retrieved
+    accumulatedResponses.forEach((item) => {
+      logEvent(
+        ability.userID,
+        {
+          type: "Response",
+          id: item.submissionID,
+        },
+        "RetrieveResponses",
+        `Retrieve selected responses for form ${formID} with ID ${item.submissionID}`
+      );
+    });
+
+    return {
+      submissions: accumulatedResponses,
+    };
+  } catch (e) {
+    logMessage.error(e);
+    return { submissions: [] };
+  }
+}
+
+/**
+ * Sets who last downloaded the Form Submission on the Vault Submission record
+ * Note that if any single update fails, the entire transaction will fail
+ * @param responses Array of responses (id, status) to update
+ * @param formID Form ID the Submission is for
+ * @param email Email address of the user downloading the Submission
+ */
+export async function updateLastDownloadedBy(
+  responses: Array<{ id: string; status: string }>,
+  formID: string,
+  email: string
+) {
+  const documentClient = connectToDynamo();
+
+  // TransactWriteItem can only update 100 items at a time
+  const asyncUpdateRequests = chunkArray(responses, 100).map((chunkedResponses) => {
+    const input: TransactWriteItemsCommandInput = {
+      TransactItems: chunkedResponses.map((response) => {
+        const isNewResponse = response.status === VaultStatus.NEW;
+        return {
+          Update: {
+            TableName: "Vault",
+            Key: {
+              FormID: {
+                S: formID,
+              },
+              NAME_OR_CONF: {
+                S: `NAME#${response.id}`,
+              },
+            },
+            UpdateExpression: "SET LastDownloadedBy = :email, DownloadedAt = :downloadedAt".concat(
+              isNewResponse ? ", #status = :statusUpdate" : ""
+            ),
+            ExpressionAttributeValues: {
+              ":email": { S: email ?? "" },
```

Comment:
According to the function signature the `email` parameter cannot be null so you should not need to `?? ""`